PERF/CI: fix benchmark import error + run asv check on all builds

pv opened the pull request:
Fix benchmark suite import error on master and 0.23.
Run `asv check` on all commits to lint the benchmark suite.

Also, fixed a timeout error that appears on newer asv versions, 
which have an updated asizeof.py module.

pv committed:
PERF: fix asv import error

pv committed:
CI: run asv check on all builds

WillAyd commented on asv_bench/benchmarks/groupby.py:
```diff
@@ -6,10 +6,15 @@
 import pandas.util.testing as tm
 
+try:
+    from pandas import TimeGrouper
```
You can remove TimeGrouper altogether (see #26477)

pv replied:
Ok.

WillAyd commented on asv_bench/benchmarks/io/parsers.py:
```diff
@@ -1,7 +1,10 @@
-from pandas._libs.tslibs.parsing import (
-    _concat_date_cols, _does_string_look_like_datetime)
+try:
+    from pandas._libs.tslibs.parsing import (
```
What was failing here? I don't think passing on exception is desired

pv replied:
The import fails with Pandas 0.23. The import errors are ignored because the benchmark suite is supposed to be importable also on historical pandas releases.

WillAyd replied:
Right but I don't think full compatibility with old versions is a goal for the benchmarking suite; that would be a nightmare to maintain! If it's not currently in master then OK to remove

pv replied:
It is currently in master.

The point is not to maintain full compatibility of the benchmark suite, but that the failures are not allowed to occur at import time.

WillAyd replied:
So IIUC you are saying this prevents any benchmark in the module from being run for historical versions where these imports aren't present?

pv replied:
Yes, it prevents running any of the benchmarks if the import fails.

That's of course maybe not optimal, it could be better if import failure disabled only the benchmarks per module, but that's how current versions of asv work.

WillAyd replied:
Hmm still a little unsure about this. What criteria would you use to judge when to do this or not?

TomAugspurger replied:
I would trust @pv on this :)

pv replied:
I would decide which pandas versions are still important to benchmark against, and then check the benchmark suite is importable with the oldest one. Here, I thought 0.23 is recent enough to warrant some adjustments.

WillAyd replied:
Gotcha. Can you just add a comment explaining why we do this then? Ideally this wouldn't prevent the entire suite from running so I think comment would be a good indicator that we should revert if asv gains that robustness in the future

pv replied:
Added.

pv committed:
PERF: since TimeGrouper was removed, remove benchmarks concerning it

pv committed:
PERF: fix benchmark frame_methods.Iteration.mem_itertuples_to_list

The runtime of the benchmark increased in asv 0.4 (which has upgraded
asv.extern.asizeof), so bump the timeout upward.

WillAyd commented:
So I see the benchmark step took 15 minutes in this PR - that would essentially run every time now right?

pv commented:
No, `asv check` takes a few seconds. The second part of the benchmark check, still runs in the same way as before this PR --- only when the benchmark suite itself has been changed.

WillAyd commented:
Makes sense - wasn't registering at first that the benchmarks here are part of the diff :-)

WillAyd reviewed (approved):
@TomAugspurger

TomAugspurger commented:
Thanks @pv!
